plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,102 +67,135 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
+                            fullyPushedDomains.add(columnHandle);
+                            return CassandraCqlUtils.validColumnName(columnHandle.getName()) + "IN (" + inValues + ")";
+                        }
+                        return translateRangeIntoCql(columnHandle, ranges.getSpan());
+                    }, discreteValues -> {
+                        if (discreteValues.isInclusive()) {
+                            if (discreteValues.getValuesCount() == 0) {
                                 return null;
                             }
-                            if (range.isSingleValue()) {
-                                singleValues.add(columnHandle.getCassandraType().toCqlLiteral(range.getSingleValue()));
+                            if (discreteValues.getValuesCount() == 1) {
+                                fullyPushedDomains.add(columnHandle);
+                                return format("%s = %s",
+                                        CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                                        toCqlLiteral(columnHandle, getOnlyElement(discreteValues.getValues())));
                             }
-                            else {
-                                if (!range.isLowUnbounded()) {
-                                    String lowBound = columnHandle.getCassandraType().toCqlLiteral(range.getLowBoundedValue());
-                                    rangeConjuncts.add(format(
-                                            "%s %s %s",
-                                            CassandraCqlUtils.validColumnName(columnHandle.getName()),
-                                            range.isLowInclusive() ? ">=" : ">",
-                                            lowBound));
-                                }
-                                if (!range.isHighUnbounded()) {
-                                    String highBound = columnHandle.getCassandraType().toCqlLiteral(range.getHighBoundedValue());
-                                    rangeConjuncts.add(format(
-                                            "%s %s %s",
-                                            CassandraCqlUtils.validColumnName(columnHandle.getName()),
-                                            range.isHighInclusive() ? "<=" : "<",
-                                            highBound));
+                            if (isInStatementNotAllowed(clusteringColumns, cassandraVersion, currentlyProcessedClusteringColumn)) {
+                                if (domain.getType().isComparable()) {
+                                    List<Range> ranges = discreteValues.getValues().stream()
+                                            .map(value -> Range.equal(domain.getType(), value))
+                                            .collect(toImmutableList());
+                                    return translateRangeIntoCql(columnHandle, SortedRangeSet.copyOf(domain.getType(), ranges).getSpan());
```

Comment:
We get here for EquatableValueSet and that is used only for types are comparable but not orderable, so taking a span of these values is probably wrong. E.g. you could try this for ColorType, Range will probably fail to create comparisonOperator from that.

Reply:
You will never get here when isOrderable is true because EquatableValueSet is used only when isOrderable is false and isComparable is true.